### PR TITLE
[#2894] Googlemaps Update Function

### PR DESCRIPTION
**Pull request by cadecairos**
Lighthouse Ticket: https://webmademovies.lighthouseapp.com/projects/65733-popcorn-maker/tickets/2894

### CODE REVIEW

File: src/core/popcorn-wrapper.js [hunk 1]
```diff
@@ -66,6 +66,7 @@ define( [ "core/logger", "core/eventmanager", "util/uri" ], function( Logger, Ev
     }; //interrupt
 
     // Update Popcorn events with data from a butter trackevent
+
```
Comment (mjschranz) [inline src/core/popcorn-wrapper.js]: Remove this.

File: src/default-config.json [hunk 2]
```diff
@@ -9,7 +9,7 @@
   "title": "Popcorn Maker",
   "ui": {
     "enabled": true,
-    "onLeaveDialog": true,
+    "onLeaveDialog": false,
```
Comment (mjschranz) [inline src/default-config.json]: Why was this changed?

File: templates/assets/editors/googlemap.js [hunk 3]
```diff
@@ -26,76 +25,68 @@ EditorHelper.addPlugin( "googlemap", function( trackEvent, popcornInstance ) {
...
+
+          updateOptions.heading = pov.heading;
+          updateOptions.pitch = pov.pitch;
+          updateOptions.zoom = pov.zoom;
+          if ( latlng ) {
+            updateOptions.lat = latlng.lat();
+            updateOptions.lng = latlng.lng();
+          }
+
+          trackEvent.update( updateOptions);
```
Comment (mjschranz) [inline templates/assets/editors/googlemap.js]: Space.

File: src/editor/trackevent-editor.js [hunk 4]
```diff
@@ -664,6 +664,9 @@ define([ "util/lang", "util/keys", "util/time", "./base-editor", "ui/widget/tool
         }
         try {
           editorElement.type = manifestEntry.type;
+          if ( editorElement.type === "number" ) {
+            editorElement.step = "any";
+          }
```
Comment (mjschranz) [inline src/editor/trackevent-editor.js]: It doesn't look like you use this code anywhere. Can you explain it's purpose?

File: templates/assets/plugins/googlemap/popcorn.googlemap.js [hunk 5]
```diff
@@ -446,6 +448,121 @@ var googleCallback;
...
+              if ( status === google.maps.GeocoderStatus.OK ) {
+                options.lat = results[ 0 ].geometry.location.lat();
+                options.lng = results[ 0 ].geometry.location.lng();
+                _cachedGeoCode[ options.location ] = location = new google.maps.LatLng( options.lat, options.lng );
+                map.panTo( location );
+              }
+            });
+          }
+        }
+
```
Comment (mjschranz) [inline templates/assets/plugins/googlemap/popcorn.googlemap.js]: Extra line.

File: templates/assets/plugins/googlemap/popcorn.googlemap.js [hunk 6]
```diff
@@ -446,6 +448,121 @@ var googleCallback;
...
+          trackEvent.top = options.top;
+          trackEvent._container.style.top = trackEvent.top + "%";
+        }
+
+        if ( options.height && options.height !== trackEvent.height ) {
+          trackEvent.height = options.height;
+          trackEvent._container.style.height = trackEvent.height + "%";
+        }
+
+        if ( options.width && options.width !== trackEvent.width ) {
```
Comment (mjschranz) [inline templates/assets/plugins/googlemap/popcorn.googlemap.js]: You do checks to ensure top/left aren't under zero. Why not the same for width/height?

File: src/default-config.json (hunk 2 above)
Comment (cadecairos) [inline src/default-config.json]: woops

File: src/editor/trackevent-editor.js (hunk 4 above)
Comment (cadecairos) [inline src/editor/trackevent-editor.js]: That fixes inputs with number types in the editor. without it any value not divisible by one is shown as invalid.

File: templates/assets/plugins/googlemap/popcorn.googlemap.js (hunk 6 above)
Comment (cadecairos) [inline templates/assets/plugins/googlemap/popcorn.googlemap.js]: I don't see why anyone would want height or width to be zero... but I can check anyways?

File: templates/assets/editors/googlemap.js [hunk 7]
```diff
@@ -26,76 +25,100 @@ EditorHelper.addPlugin( "googlemap", function( trackEvent, popcornInstance ) {
...
+            streetViewTimeout,
+            zoomTimeout;
+
+        function locationChanged() {
+
+          if ( locationTimeout ) {
+            clearTimeout( locationTimeout );
+          }
+
+          setTimeout(function() {
```
Comment (mjschranz) [inline templates/assets/editors/googlemap.js]: This should be setting the timeouts id to `locationTimeout`.

File: templates/assets/editors/googlemap.js [hunk 8]
```diff
@@ -26,76 +25,108 @@ EditorHelper.addPlugin( "googlemap", function( trackEvent, popcornInstance ) {
...
-    if ( !trackEvent.popcornTrackEvent._map ) {
-      trackEvent.popcornTrackEvent.onmaploaded = function( options, map ) {
-        popcornEventMapReference = map;
+      if ( !trackEvent.popcornTrackEvent._map ) {
+        trackEvent.popcornTrackEvent.onmaploaded = function( options, map ) {
+          popcornEventMapReference = map;
+          setupMapListeners();
+        };
+      }
+      else {
```
Comment (mjschranz) [inline templates/assets/editors/googlemap.js]: Small nit. I know that it was like this before but ideally we could bump this onto the line above so we gradually keep our code style in check.

File: src/editor/trackevent-editor.js (hunk 4 above)
Comment (jbuck) [inline src/editor/trackevent-editor.js]: So this basically handles the `0` case? If so, please add a comment to that effect because it's not obvious at all

Comment (cadecairos) [inline src/editor/trackevent-editor.js]: The specific reason for this case was heading and pitch in Google Maps. they can be set to numbers like `23.812648112311` and without the `step="any"` chrome defaults to `step="1"`, which only allows for "whole" numbers like 1, 2, 3..

Comment (cadecairos) [inline src/editor/trackevent-editor.js]: Might be worth noting that the step attribute isn't implemented in any other browsers. Maybe in Opera, though I'm not entirely sure.

File: templates/assets/editors/googlemap.js [hunk 9]
```diff
@@ -1,7 +1,6 @@
-/*global EditorHelper,google,clearTimeout,setTimeout*/
+/*global EditorHelper,google*/
 
 EditorHelper.addPlugin( "googlemap", function( trackEvent, popcornInstance ) {
```
Comment (jbuck) [inline templates/assets/editors/googlemap.js]: Instead of doing `var popcorn = popcornInstance`, change this variable name to `popcorn`

File: templates/assets/editors/googlemap.js (hunk 8 above)
Comment (jbuck) [inline templates/assets/editors/googlemap.js]: Yep, it'd be nice to fix this

File: templates/assets/editors/googlemap.js [hunk 10]
```diff
@@ -26,76 +25,108 @@ EditorHelper.addPlugin( "googlemap", function( trackEvent, popcornInstance ) {
...
-      setup();
-    }
+        setup();
+      }
 
-    if ( !trackEvent.popcornTrackEvent._map ) {
-      trackEvent.popcornTrackEvent.onmaploaded = function( options, map ) {
-        popcornEventMapReference = map;
+      if ( !trackEvent.popcornTrackEvent._map ) {
+        trackEvent.popcornTrackEvent.onmaploaded = function( options, map ) {
```
Comment (jbuck) [inline templates/assets/editors/googlemap.js]: Maybe I'm reading this wrong, but aren't you already listening for this with `popcorn.on( "googlemaps-loaded")`, so this is actually a dead code path?

File: templates/assets/plugins/googlemap/popcorn.googlemap.js [hunk 11]
```diff
@@ -217,19 +184,63 @@ var googleCallback;
...
+
+        options.toString = function() {
+          if ( options.type === "STREETVIEW" ) {
+            return "Streeview";
+          }
+          return options.location || ( ( options.lat && options.lng ) ? options.lat + ", " + options.lng : options._natives.manifest.options.location[ "default" ] );
+        };
+      },
       /**
        * @member webpage
```
Comment (jbuck) [inline templates/assets/plugins/googlemap/popcorn.googlemap.js]: what is this i don't even
